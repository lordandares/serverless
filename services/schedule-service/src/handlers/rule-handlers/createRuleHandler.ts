import { errors, schemas } from '@lighthouse/serverless-common'
import { Context } from 'aws-lambda'
import * as AWS from 'aws-sdk'
import { ClientConfiguration, DocumentClient } from 'aws-sdk/clients/dynamodb'
import * as AWSXRay from 'aws-xray-sdk'
import * as moment from 'moment'

import { OCCURRENCE_STATUS_ACTIVE } from '../../constants'
import { payloadToRulePatternDocument } from '../../service/lib/ruleTransform'

declare var process: {
  env: {
    IS_OFFLINE: string
    REGION: string
    OCCURRENCE_ACTIVE_ARN: string
    TABLE_SCHEDULES: string
  }
}

export async function createRuleHandler(body: any, context: Context) {
  try {
    const { awsRequestId: arnId, invokedFunctionArn: arn } = context

    if (!body) {
      throw new errors.ApplicationError({
        message: `A rule body is required`,
      })
    }

    await schemas.validate({
      data: body,
      schema: schemas.rulePayloadSchema,
    })

    console.info('CreateRuleHandler: body', { body })

    const isOffline = process.env.IS_OFFLINE
    const region = process.env.REGION
    const tableName = process.env.TABLE_SCHEDULES

    const dynamoDbConfig: ClientConfiguration =
      region === 'localhost'
        ? /* istanbul ignore next */ {
            endpoint: 'http://localhost:8000',
            httpOptions: { timeout: 5000 },
            region,
          }
        : {}

    const DynamoDB = isOffline
      ? AWS.DynamoDB
      : /* istanbul ignore next */
        AWSXRay.captureAWS(AWS).DynamoDB

    const documentClient: DocumentClient = new DynamoDB.DocumentClient(
      dynamoDbConfig,
    )

    const {
      applicationId,
      locationId,
      occurrenceId,
      pk: occurrencePk,
      scheduleId,
      startAt: startAtIsoString,
      sk: occurrenceSk,
    } = body

    const nowIsoString = new Date().toISOString()
    const isStartInFuture = startAtIsoString > nowIsoString

    // NOTE: if the start is in the future we prevent the rule from being created
    // until one minute before the occurrence is due to start
    if (isStartInFuture) {
      console.info(
        'CreateRuleHandler: occurrence is in the future so creating timer',
        {
          nowIsoString,
          startAtIsoString,
        },
      )

      const mNow = moment(new Date(startAtIsoString))

      const timerTriggersAt = mNow.utc().subtract(1, 'minute')
      const timerTriggersAtMin = timerTriggersAt.format('YYYY-MM-DDTHH:mm')

      const timerExpiresAt = mNow
        .add(1, 'day')
        .toDate()
        .toISOString()

      const timerDocument: schemas.TimerDocumentSchema = {
        expiresAt: timerExpiresAt,
        groupType: 'occurrence-timer',
        pk: `timer-${timerTriggersAtMin}`,
        scheduleId,
        sk: `timer#${occurrencePk}#${occurrenceSk}`,
        targetArn: process.env.OCCURRENCE_ACTIVE_ARN,
      }

      await schemas.validate({
        data: timerDocument,
        schema: schemas.timerDocumentSchema,
      })

      const putTimerParams: DocumentClient.PutItemInput = {
        Item: timerDocument,
        TableName: tableName,
      }

      console.info('CreateRuleHandler: creating timer', {
        timerDocument,
      })

      await documentClient.put(putTimerParams).promise()

      console.info('CreateRuleHandler: created timer')

      return
    }

    const pk = `rule-pattern-${applicationId}-${locationId}`
    const sk = 'visit'

    const rulePatternGetParams: DocumentClient.GetItemInput = {
      Key: { pk, sk },
      TableName: tableName,
    }

    const rulePatternGetResponse: DocumentClient.GetItemOutput = await documentClient
      .get(rulePatternGetParams)
      .promise()

    const { Item: rulePatternDocument } = rulePatternGetResponse

    const actor: schemas.ActorSchema = {
      id: arnId,
      label: arn,
      type: 'system',
    }

    const patternDocument: schemas.RulePatternDocumentSchema = payloadToRulePatternDocument(
      {
        actor,
        existingDocument: rulePatternDocument as schemas.RulePatternDocumentSchema,
        payload: body,
      },
    )
    try {
      await schemas.validate({
        data: patternDocument,
        schema: schemas.rulePatternDocumentSchema,
      })
    } catch (err) {
      throw new errors.ApplicationError({ message: err.message })
    }

    const occurrenceGetParams: DocumentClient.GetItemInput = {
      Key: { pk: occurrencePk, sk: occurrenceSk },
      TableName: tableName,
    }

    const occurrenceGetResponse: DocumentClient.GetItemOutput = await documentClient
      .get(occurrenceGetParams)
      .promise()

    const { Item: occurrenceDocument } = occurrenceGetResponse

    if (!occurrenceDocument) {
      throw new errors.ApplicationError({
        message: `The occurrence is missing so unable to add rule or set occurrence to active`,
      })
    }

    const transactItems: DocumentClient.TransactWriteItemList = []

    const updatedAtIsoString = new Date().toISOString()

    // update or add new rule pattern document
    if (rulePatternDocument) {
      console.info(
        'CreateRuleHandler: updating existing rule pattern document',
        {
          patternDocument,
        },
      )

      transactItems.push({
        Update: {
          ConditionExpression: '#updatedAt = :expectedUpdatedAt',
          ExpressionAttributeNames: {
            '#matches': 'matches',
            '#updatedAt': 'updatedAt',
            '#updatedBy': 'updatedBy',
          },
          ExpressionAttributeValues: {
            ':expectedUpdatedAt': rulePatternDocument.updatedAt,
            ':matches': patternDocument.matches,
            ':nextUpdatedAt': updatedAtIsoString,
            ':updatedBy': actor,
          },
          Key: { pk: patternDocument.pk, sk: patternDocument.sk },
          TableName: tableName,
          UpdateExpression:
            'SET #matches = :matches, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy',
        },
      })
    } else {
      console.info('CreateRuleHandler: adding new rule pattern document', {
        patternDocument,
      })

      transactItems.push({
        Put: {
          Item: patternDocument,
          TableName: tableName,
        },
      })
    }

    console.info('CreateRuleHandler: updating existing occurrence', {
      occurrenceDocument,
    })

    // update occurrence so status is now active
    transactItems.push({
      Update: {
        ConditionExpression: '#updatedAt = :expectedUpdatedAt',
        ExpressionAttributeNames: {
          '#status': 'status',
          '#updatedAt': 'updatedAt',
          '#updatedBy': 'updatedBy',
        },
        ExpressionAttributeValues: {
          ':expectedUpdatedAt': occurrenceDocument.updatedAt,
          ':nextUpdatedAt': updatedAtIsoString,
          ':status': OCCURRENCE_STATUS_ACTIVE,
          ':updatedBy': actor,
        },
        Key: { pk: occurrenceDocument.pk, sk: occurrenceDocument.sk },
        TableName: tableName,
        UpdateExpression:
          'SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy',
      },
    })

    const transactParams: DocumentClient.TransactWriteItemsInput = {
      TransactItems: transactItems,
    }

    const transactItemsLength = transactItems.length

    try {
      console.info(
        `CreateRuleHandler: processing ${transactItemsLength} transactions`,
      )

      await documentClient.transactWrite(transactParams).promise()

      console.info('CreateRuleHandler: completed transactions successfully')
    } catch (err) {
      throw new errors.ApplicationError({ message: err.message })
    }
  } catch (err) {
    console.error('CreateRuleHandlerError', {
      body,
      err,
    })

    throw err
  }
}
