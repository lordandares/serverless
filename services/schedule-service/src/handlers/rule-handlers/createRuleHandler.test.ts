import { schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'

import * as MockDate from 'mockdate'
import {
  context,
  occurrenceDocument,
  rulePatternDocument,
  rulePayload,
} from '../../__fixtures__'
import { createRuleHandler } from './createRuleHandler'

jest.mock('uuid/v4', () => () => 'action-uuid-1234')

beforeEach(() => {
  process.env.IS_OFFLINE = 'true'
  process.env.OCCURRENCE_ACTIVE_ARN = 'occurrence-active-arn'
  process.env.TABLE_SCHEDULES = 'table-schedules'
  MockDate.set('2019-02-01T01:00:00.000Z')
})

afterEach(() => MockDate.reset())

test('application error when body missing', async () => {
  expect.assertions(1)

  try {
    await createRuleHandler(null, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: A rule body is required]`,
    )
  }
})

test('validates against schema', async () => {
  expect.assertions(1)

  const validateSpy = jest.spyOn(schemas, 'validate')
  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: undefined })
    .mockResolvedValueOnce({ Item: occurrenceDocument })
  const transactSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  await createRuleHandler(rulePayload, context)

  expect(validateSpy).toHaveBeenCalledWith({
    data: rulePayload,
    schema: schemas.rulePayloadSchema,
  })

  validateSpy.mockRestore()

  AWS.restore()
})

test('creates timer document when occurrence starts in the future', async () => {
  expect.assertions(1)

  const putSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)

  const payload = { ...rulePayload, startAt: '2020-01-01T00:00:00.000Z' }
  await createRuleHandler(payload, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  AWS.restore()
})

test('creates new rule pattern document when none existing', async () => {
  expect.assertions(2)

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: undefined })
    .mockResolvedValueOnce({ Item: occurrenceDocument })
  const transactSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  await createRuleHandler(rulePayload, context)

  expect(getSpy.mock.calls).toMatchSnapshot()
  expect(transactSpy.mock.calls).toMatchSnapshot()

  AWS.restore()
})

test('updates existing rule pattern document when one exists', async () => {
  expect.assertions(2)

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: rulePatternDocument })
    .mockResolvedValueOnce({ Item: occurrenceDocument })
  const transactSpy = jest.fn().mockResolvedValue({})

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  await createRuleHandler(rulePayload, context)

  expect(getSpy.mock.calls).toMatchSnapshot()
  expect(transactSpy.mock.calls).toMatchSnapshot()

  AWS.restore()
})

test('application error when occurrence missing', async () => {
  expect.assertions(1)

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: rulePatternDocument })
    .mockResolvedValueOnce({ Item: undefined })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  try {
    await createRuleHandler(rulePayload, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The occurrence is missing so unable to add rule or set occurrence to active]`,
    )
  }

  AWS.restore()
})

test('application error when transactions fail', async () => {
  expect.assertions(1)

  const getSpy = jest
    .fn()
    .mockResolvedValueOnce({ Item: rulePatternDocument })
    .mockResolvedValueOnce({ Item: occurrenceDocument })
  const transactSpy = jest
    .fn()
    .mockRejectedValue(new Error('Transaction Error!'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactSpy)

  try {
    await createRuleHandler(rulePayload, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: Transaction Error!]`,
    )
  }

  AWS.restore()
})

test('patternDocument validation error', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: undefined })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  schemas.validate = jest
    .fn()
    .mockResolvedValueOnce({})
    .mockRejectedValue(new Error('Validation Error 2'))

  try {
    await createRuleHandler(rulePayload, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: Validation Error 2]`,
    )
  }

  AWS.restore()
})
