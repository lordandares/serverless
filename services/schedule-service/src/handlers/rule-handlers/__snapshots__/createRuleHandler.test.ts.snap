// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates new rule pattern document when none existing 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "pk": "rule-pattern-application-id-1111-location-id-11111",
        "sk": "visit",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Key": Object {
        "pk": "application1-occurrence",
        "sk": "2019-06-02T00:00:00.000Z-occurrence1",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`creates new rule pattern document when none existing 2`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Put": Object {
            "Item": Object {
              "createdAt": "2019-02-01T01:00:00.000Z",
              "createdBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
              "groupType": "rule-pattern",
              "matches": Object {
                "occurrence1": Object {
                  "pk": "application1-occurrence",
                  "sk": "2019-06-02T00:00:00.000Z-occurrence1",
                },
              },
              "pk": "rule-pattern-application-id-1111-location-id-11111",
              "sk": "visit",
              "updatedAt": "2019-02-01T01:00:00.000Z",
              "updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "TableName": "table-schedules",
          },
        },
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#status": "status",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": undefined,
              ":nextUpdatedAt": "2019-02-01T01:00:00.000Z",
              ":status": "active",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "application1-occurrence",
              "sk": "2019-01-01T00:00-occurrence1",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
      ],
    },
    [Function],
  ],
]
`;

exports[`creates timer document when occurrence starts in the future 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "expiresAt": "2020-01-01T23:59:00.000Z",
        "groupType": "occurrence-timer",
        "pk": "timer-2019-12-31T23:59",
        "scheduleId": "schedule1",
        "sk": "timer#application1-occurrence#2019-06-02T00:00:00.000Z-occurrence1",
        "targetArn": "occurrence-active-arn",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`updates existing rule pattern document when one exists 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "pk": "rule-pattern-application-id-1111-location-id-11111",
        "sk": "visit",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Key": Object {
        "pk": "application1-occurrence",
        "sk": "2019-06-02T00:00:00.000Z-occurrence1",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`updates existing rule pattern document when one exists 2`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#matches": "matches",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": "2019-06-02T00:00:00.000Z",
              ":matches": Object {
                "occurrence1": Object {
                  "pk": "application1-occurrence",
                  "sk": "2019-06-02T00:00:00.000Z-occurrence1",
                },
              },
              ":nextUpdatedAt": "2019-02-01T01:00:00.000Z",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "rule-pattern-application-id-1111-location-id-11111",
              "sk": "visit",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #matches = :matches, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
        Object {
          "Update": Object {
            "ConditionExpression": "#updatedAt = :expectedUpdatedAt",
            "ExpressionAttributeNames": Object {
              "#status": "status",
              "#updatedAt": "updatedAt",
              "#updatedBy": "updatedBy",
            },
            "ExpressionAttributeValues": Object {
              ":expectedUpdatedAt": undefined,
              ":nextUpdatedAt": "2019-02-01T01:00:00.000Z",
              ":status": "active",
              ":updatedBy": Object {
                "id": "aws-request-id",
                "label": "invoked-function-arn",
                "type": "system",
              },
            },
            "Key": Object {
              "pk": "application1-occurrence",
              "sk": "2019-01-01T00:00-occurrence1",
            },
            "TableName": "table-schedules",
            "UpdateExpression": "SET #status = :status, #updatedAt = :nextUpdatedAt, #updatedBy = :updatedBy",
          },
        },
      ],
    },
    [Function],
  ],
]
`;
